File: meetupCpp/bench.cpp

#include <vector>
#include <string>
#include <fstream>
#include <iostream>
#include <chrono>
#include <algorithm>
#include <cstring>

using namespace std;

static const std::size_t SIZE = 128;

class LineBuf{
public:
    LineBuf(string s = ""){
        std::memcpy(buf, s.c_str(), (s.size()<SIZE)?s.size():SIZE);
    }
    char buf[SIZE] = {0};

    bool operator < (LineBuf& rval){
        return std::memcmp(buf, rval.buf, SIZE) < 0;
    }
};

template<class T>
void lines_from_file(vector<T> &lines, string filename){

    ifstream file(filename.c_str());
    string line;
    while(getline(file, line)){
        for(size_t pos=0; pos<line.size(); pos+= SIZE){
            auto str = line.substr(pos, SIZE);
            lines.emplace_back(str);
        }

    }
}

template<class T>
void bench(vector<T> &lines){
    cout<<"Read: "<< lines.size() <<" lines"<<endl;

    auto begin = chrono::steady_clock::now();
    sort(lines.begin(), lines.end());
    auto end = chrono::steady_clock::now();

    auto elapsed_ms = chrono::duration_cast<chrono::milliseconds>(end - begin);
    cout << "The time: " << elapsed_ms.count() << " ms\n";
}

int main(){
    cout<<"Test vector<strings>:"<<endl;
    vector<string> lines_str;
    lines_from_file<string>(lines_str, "/home/alexey/tmp/data.bin");
    bench<string>(lines_str);

    cout<<"Test vector<char["<<SIZE<<"]>:"<<endl;
    vector<LineBuf> lines_custom;
    lines_from_file<LineBuf>(lines_custom, "/home/alexey/tmp/data.bin");
    bench<LineBuf>(lines_custom);

    return 0;
}

File: meetupRust/src/bench.rs
use std::time::Instant;

use std::{
    fs::File,
    io::{prelude::*, BufReader},
    path::Path,
};
use std::cmp::Ordering;

const SIZE : usize = 256;

struct LineBuf{
    buf: [char; SIZE]
}

impl Default for LineBuf{
    fn default() -> LineBuf {
        LineBuf {
            buf: ['\0'; SIZE]
        }
    }
}

impl Clone for LineBuf {
    fn clone(&self) -> Self {
        LineBuf {
            buf: self.buf.clone()
        }
    }
}

fn lines_from_file(filename: impl AsRef<Path>) -> (Vec<String>, Vec<LineBuf>) {
    let file = File::open(filename).expect("no such file");
    let buf = BufReader::new(file);
    let lines :Vec<String> =buf.lines()
        .map(|l| l.expect("Could not parse line"))
        .collect();

    let mut str_vec = vec![];
    let mut buf_vec = vec![];

    for line in lines.iter(){
        for idx in 0..(line.len() / SIZE){
            let sub_line = &line[idx*SIZE .. (idx+1)*SIZE].to_string();

            str_vec.push(sub_line.clone());
            let mut buf_line : LineBuf = LineBuf {buf: ['\0'; SIZE]};
            let mut pos = 0;
            for elem in sub_line.chars(){
                buf_line.buf[pos] = elem;
                pos += 1;
            }

            buf_vec.push(buf_line.clone());
        }
    }

    (str_vec, buf_vec)
}



fn main() {
    let ( mut lines_str, mut lines_buf) = lines_from_file("/home/alexey/tmp/data.bin");

    println!("Vec<String> Read: {} lines", lines_str.len());
    let now = Instant::now();
    lines_str.sort();
    let elapsed = now.elapsed().as_millis();
    println!("Sorted: {} ms", elapsed);

    println!("Vec<u8[{}]> Read: {} lines", SIZE, lines_buf.len());
    let now_buf = Instant::now();
    lines_buf.sort_by(|a, b| a.buf.cmp(&b.buf));
    let elapsed_buf = now_buf.elapsed().as_millis();
    println!("Sorted: {} ms", elapsed_buf);

}